### PR TITLE
Keep interrupt priority group value unchanged when writing to the AIRCR register

### PR DESCRIPTION
**Pull request by mmahadevan108**
The issue was uncovered when issuing a SYSTEM RESET on the i.MXRT platform.

Signed-off-by: Mahesh Mahadevan <Mahesh.Mahadevan@nxp.com>

### CODE REVIEW

**Comment by mmahadevan108** [comment]
cc @c1728p9 @flit

**Comment by flit** [comment]
What was the symptom of this bug? I'm actually surprised `PRIGROUP` doesn't get reset automaticaly by the core when you perform the system reset.

**Comment by mmahadevan108** [comment]
We would see a FAIL.TXT file with the message "The interface firmware FAILED to reset/halt the target MCU"

**Comment by flit** [comment]
Do you know what the root cause was? Is it simply that you can't change any other fields of AIRCR when setting `VECTRESET`?

**Comment by mmahadevan108** [comment]
The failure to reset/halt the target MCU is seen when using SYSTEMRESET without the patch to retain interrupt priority bits.

**Comment by c1728p9** [comment]
The priority bits should get reset when a software reset occurs. This is from the ARMv7m architecture reference manual:
```
PRIGROUP, bits[10:8]
Priority grouping, indicates the binary point position.
For information about the use of this field see Priority grouping on page B1-583.
This field resets to 0b000.
```

Can you confirm with a debugger or logging code in DAPLink that these bits aren't getting reset?

**Comment by flit** [comment]
The change is not because `PRIGROUP` isn't getting reset, it is to perform a read-modify-write so that only `VECTRESET` is changed in the write to AIRCR. After the core resets, `PRIGROUP` should read as zeros.

I think the issue here is that you are not supposed to modify any other writable bits in the AIRCR when you are setting `VECTRESET` or the behaviour is undefined. The v7-M ARM is not very clear about this, but it makes sense given the way some of the other registers work. Also, this may be different on the CM7 versus other cores.

**Comment by c1728p9** [comment]
@flit this patch only preserves the bits in SCB_AIRCR_PRIGROUP_Msk, the others are set to 0 as they were before. Because of this, I'm a little surprised this fixes the issue.

**Comment by mmahadevan108** [comment]
The CMSIS code for System Reset has a similar implementation:
https://github.com/ARMmbed/mbed-os/blob/master/cmsis/TARGET_CORTEX_M/core_cm7.h

/**
  \brief   System Reset
  \details Initiates a system reset request to reset the MCU.
 */
__STATIC_INLINE void __NVIC_SystemReset(void)
{
  __DSB();                                                          /* Ensure all outstanding memory accesses included
                                                                       buffered write are completed before reset */
  SCB->AIRCR  = (uint32_t)((0x5FAUL << SCB_AIRCR_VECTKEY_Pos)    |
                           (SCB->AIRCR & SCB_AIRCR_PRIGROUP_Msk) |
                            SCB_AIRCR_SYSRESETREQ_Msk    );         /* Keep priority group unchanged */
  __DSB();                                                          /* Ensure completion of memory access */

  for(;;)                                                           /* wait until reset */
  {
    __NOP();
  }
}

**Comment by mmahadevan108** [comment]
cc @maclobdell

**Comment by mmahadevan108** [comment]
Can this be merged?

**Comment by c1728p9** [comment]
Hi @mmahadevan108 this fix needs to be understood and verified before it can be merged. Can you confirm with a debugger or logging code in DAPLink that these bits aren't getting reset? Alternatively, if you can provide me with the steps to reproduce this then I can try and replicate this locally. If it can be confirmed that this behavior in not to spec then I can escalate this issue.

**Comment by mmahadevan108** [comment]
The reset code implementation was changed to match what is present in the CMSIS code. Could you help check with the team in ARM that provides this.

**Comment by c1728p9** [comment]
I opened an issue in the CMSIS_5 repo to find out more information. In the mean time could you either confirm this locally or provide me with steps to reproduce the issue?

**Comment by flit** [comment]
@c1728p9 Can this PR be merged so the reset code is aligned with CMSIS_5?

To respond to an earlier comment from you, PRIGROUP is the only field of AIRCR that is r/w. Thus it is the only field that would need to be preserved in a read-modify-write.

**Comment by c1728p9** [comment]
Hi @mmahadevan108 this isn't a failure due to this PR. I'm working to get CI fixed.

**Comment by c1728p9** [comment]
@mmahadevan108 this time the build failure is legitimate. If you try to build kl26z_microbit_if locally you should see these build failures.

**Comment by mmahadevan108** [comment]
Thank you, the priority group bits do not exist on Cortex-M0 & Cortex-M0+. I have added a check to account for this.

File: source/daplink/interface/swd_host.c [hunk 1]
```diff
@@ -1005,10 +1005,19 @@ uint8_t swd_set_target_state_sw(TARGET_RESET_STATE state)
             }
 
             // Perform a soft reset
-            if (!swd_write_word(NVIC_AIRCR, VECTKEY | SOFT_RESET)) {
+#if (__CORTEX_M >= 0x03)
```
Comment (c1728p9) [inline source/daplink/interface/swd_host.c]: This define is for the MCU running DAPLink when really it depends on the target MCU. Could you define ```SCB_AIRCR_PRIGROUP_Msk``` at the top of this file so it is present for all MCUs?

**Comment by c1728p9** [comment]
@mmahadevan108 could you update this PR again?

**Pull request merged**